src/controller_examples/rabbitmq_controller/proof/helper_invariants/validation.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use crate::external_api::spec::{EmptyAPI, EmptyTypeView};
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, config_map::*, dynamic::*, error::*, owner_reference::*, resource::*,
+    stateful_set::*,
+};
+use crate::kubernetes_cluster::spec::{
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::rabbitmq_controller::{
+    common::*,
+    proof::{
+        helper_invariants::{owner_ref::*, predicate::*, proof::*},
+        predicate::*,
+        resource::*,
+    },
+    spec::{reconciler::*, resource::*, types::*},
+};
+use crate::reconciler::spec::{reconciler::*, resource_builder::*};
+use crate::temporal_logic::{defs::*, rules::*};
+use crate::vstd_ext::{multiset_lib, seq_lib, string_view::*};
+use vstd::{multiset::*, prelude::*, string::*};
+
+verus! {
+
+pub open spec fn stateful_set_satisfies_unchangeable(obj: DynamicObjectView, rabbitmq: RabbitmqClusterView) -> bool {
+    let made_spec = make_stateful_set(rabbitmq, new_strlit("")@).spec.get_Some_0();
+    let sts = StatefulSetView::unmarshal(obj).get_Ok_0();
+
+    obj.metadata.owner_references_only_contains(rabbitmq.controller_owner_ref()) ==> made_spec == StatefulSetSpecView {
+        replicas: made_spec.replicas,
+        template: made_spec.template,
+        persistent_volume_claim_retention_policy: made_spec.persistent_volume_claim_retention_policy,
+        ..sts.spec.get_Some_0()
+    }
+}
+
+pub open spec fn stateful_set_in_etcd_satisfies_unchangeable(rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    let key = rabbitmq.object_ref();
+    let sts_key = StatefulSetBuilder::get_request(rabbitmq).key;
+    |s: RMQCluster| {
+        s.resources().contains_key(key)
+        && s.resources().contains_key(sts_key)
+        ==> stateful_set_satisfies_unchangeable(s.resources()[sts_key], RabbitmqClusterView::unmarshal(s.resources()[key]).get_Ok_0())
+    }
+}
+
+pub proof fn lemma_always_stateful_set_in_etcd_satisfies_unchangeable(spec: TempPred<RMQCluster>, rabbitmq: RabbitmqClusterView)
```

Comment:
Is this lemma used anywhere (I might miss something)?

Reply:
https://github.com/vmware-research/verifiable-controllers/blob/cc3a29a71b6a0bb49bf335de022eb0d824cd6331/src/controller_examples/rabbitmq_controller/proof/liveness/spec.rs#L434

Reply:
Can you also document the lemmas here like the one in unchangeable.rs? From the name, I guess the lemma is for the validation rule but I am not sure. Also I am surprised by the removal of this https://github.com/vmware-research/verifiable-controllers/pull/340/files#r1354146855. Can you comment on it?

Reply:
I use the `unchangeable` spec to describe what properties hold all through the executions. But the things that I deleted are not like that.

Reply:
This lemma is to show that some fields of stateful set can be determined by that cr object its owner reference points to. It can be used to show the state of stateful set finally always matches the `rabbitmq` because certain fields are not updated by `update` method.

Reply:
Since those fields stay unchanged, this invariant also helps to show the transition validation must be satisfied when proving `AtStep(Update, StatefulSet) ~> resource_match` (otherwise that update request will be rejected).

Reply:
> I use the unchangeable spec to describe what properties hold all through the executions. But the things that I deleted are not like that so I removed that.

But the spec you deleted actually said that the fields (other than replicas, template and retention policy) are always the same as the ones from make_stateful_set through the executions. Note that I am not against your change, I am just trying to understand your rationale behind the change so I can better take over.

Reply:
> But the spec you deleted actually said that the fields (other than replicas, template and retention policy) are always the same as the ones from make_stateful_set through the executions.

That doesn't hold all the time so I deleted them

Reply:
> This lemma is to show that some fields of stateful set can be determined by that cr object its owner reference points to.

My confusion is mainly because I cannot really tell why we need such lemmas for the stateful set, but not for other sub-resources -- or did I miss something major here? Could you document the reason?

Reply:
Here is my guess: there are three types of fields in terms of how they get (un)changed: (1) the fields whose values are fixed and immutable for whatever owner cr, (2) the fields whose values are decided by the owner cr but still immutable since created, (3) the fields whose values are mutable.

The unchangeable spec fn talks about (1), the lemmas in this file talk about (2) (and only stateful set has such fields), and (3) is handled when you apply wf1 (which is the easy case). Is that correct?

Reply:
You are generally correct. I want to say we have the stateful set lemma is not only because of (2) but also because these fields won't be updated by the update methods.

Reply:
I update the comments.

> (1) the fields whose values are fixed and immutable for whatever owner cr,
I also mentioned this in the comments of file `unchangeable`.

Reply:
> also because these fields won't be updated by the update methods

Both the unchangeable lemma and the validation lemma here are designed to deal with this "partial update" behavior, which is fundamentally different from the early proof (where update always updates the entire object so wf1 solves all the problems regarding updates). This seems to be one of the major **new** challenges in the current proof, is my understanding correct?

Reply:
Yes, one of. But I didn't make a summary to all the new changes.

Reply:
It would be very helpful if you can describe the major new changes to the proof (compared to the old first edition) in the description of this PR. I believe the reasoning on the cm's rv number is also one of them. You can just list the major new challenges at high level, and point to the lemma that solves the problem.

I also see that you made some progress on automating the previously tedious process of unpacking phased invariants. It would be great if you can also briefly document such in the description.

Reply:
After that I will merge this PR (unless you have more to commit)